lib/Dialect/SV/Transforms/RTLLegalizeNames.cpp

```diff
@@ -0,0 +1,198 @@
+  root.getBodyRegion().walk([&](Operation *op) {
+    if (auto extMod = dyn_cast<RTLModuleExternOp>(op)) {
+      auto verilogName = extMod.getVerilogModuleNameAttr();
+      getModuleName(verilogName);
+      // We expressely *do not* rename extern modules, where the user supplies
+      // the complementary definition from the outside.
+      // TODO: Maybe we should emit a diagnostic if this pass would rename an
+      // external module.
+    }
+  });
+
+  // Rename modules and interfaces.
+  mlir::SymbolTableCollection symbolTable;
+  mlir::SymbolUserMap symbolUsers(symbolTable, root);
+  root.getBodyRegion().walk([&](Operation *op) {
+    StringAttr nameAttr;
+    if (auto module = dyn_cast<RTLModuleOp>(op)) {
+      nameAttr = module.getNameAttr();
+    } else if (auto intf = dyn_cast<InterfaceOp>(op)) {
+      nameAttr = intf.getNameAttr();
+    }
+    if (nameAttr) {
+      auto oldName = nameAttr.getValue();
+      auto newName = getModuleName(nameAttr);
+      if (oldName != newName) {
+        symbolUsers.replaceAllUsesWith(op, newName);
+        SymbolTable::setSymbolName(op, newName);
+      }
+      anythingChanged = true;
+    }
+  });
+
+  // Tables to track the used names within a given module or interface.
+  // llvm::DenseMap<ValueOrOp, StringRef> nameTable;
+  llvm::StringSet<> usedNames = usedModuleNames;
+  size_t nextGeneratedNameID = 0;
+
+  // Rename individual operations.
+  root.getBodyRegion().walk([&](Operation *op) {
+    if (auto module = dyn_cast<RTLModuleOp>(op)) {
+      auto moduleType = rtl::getModuleType(module);
+      auto inputs = moduleType.getInputs();
+      auto results = moduleType.getResults();
+
+      // Rename the inputs.
+      for (unsigned i = 0; i < inputs.size(); ++i) {
+        auto oldName = module.getArgAttrOfType<StringAttr>(i, "rtl.name");
+        if (!oldName)
+          continue;
+        auto newName =
+            sanitizeName(oldName.getValue(), usedNames, nextGeneratedNameID);
+        if (oldName.getValue() != newName) {
+          module.setArgAttr(i, "rtl.name",
+                            StringAttr::get(module.getContext(), newName));
+          anythingChanged = true;
+        }
+      }
+
+      // Rename the results.
+      for (unsigned i = 0; i < results.size(); ++i) {
+        auto oldName = module.getResultAttrOfType<StringAttr>(i, "rtl.name");
+        if (!oldName)
+          continue;
+        auto newName =
+            sanitizeName(oldName.getValue(), usedNames, nextGeneratedNameID);
+        if (oldName.getValue() != newName) {
+          module.setResultAttr(i, "rtl.name",
+                               StringAttr::get(module.getContext(), newName));
+          anythingChanged = true;
+        }
+      }
+
+      // Since `walk` does a depth-first traversal, we can clear out the used
+      // names at this point for use in the next operation.
+      usedNames = usedModuleNames;
+      nextGeneratedNameID = 0;
+    }
+
+    else if (isa<InterfaceOp>(op)) {
```

Comment:
Nit: prefer early return (out of the previous block) and use a regular `if` here.